libcontainer/rootfs_linux.go

```diff
@@ -1068,19 +1077,58 @@ func pivotRoot(rootfs string) error {
 	}
 	defer unix.Close(oldroot) //nolint: errcheck
 
-	newroot, err := unix.Open(rootfs, unix.O_DIRECTORY|unix.O_RDONLY, 0)
-	if err != nil {
-		return &os.PathError{Op: "open", Path: rootfs, Err: err}
-	}
-	defer unix.Close(newroot) //nolint: errcheck
-
 	// Change to the new root so that the pivot_root actually acts on it.
-	if err := unix.Fchdir(newroot); err != nil {
-		return &os.PathError{Op: "fchdir", Path: "fd " + strconv.Itoa(newroot), Err: err}
+	if err := os.Chdir(rootfs); err != nil {
+		return err
 	}
 
-	if err := unix.PivotRoot(".", "."); err != nil {
-		return &os.PathError{Op: "pivot_root", Path: ".", Err: err}
+	pivotErr := unix.PivotRoot(".", ".")
+	if errors.Is(pivotErr, unix.EINVAL) {
+		// If pivot_root(2) failed with -EINVAL, one of the possible reasons is
```

Comment:
There are six reasons will cause `-EINVAL`:
```
       EINVAL new_root is not a mount point.

       EINVAL put_old is not at or underneath new_root.

       EINVAL The current root directory is not a mount point (because
              of an earlier [chroot(2)](https://man7.org/linux/man-pages/man2/chroot.2.html)).

       EINVAL The current root is on the rootfs (initial ramfs) mount;
              see NOTES.

       EINVAL Either the mount point at new_root, or the parent mount of
              that mount point, has propagation type MS_SHARED.

       EINVAL put_old is a mount point and has the propagation type
              MS_SHARED.
```

This PR tries to deal with the forth one, does this pr would cause some unexpected behaviors for the fifth one? (Maybe need to check once have a time)
The user case:
For docker, containerd, kubernetes, the rootfs is always mount as a overlay fs, it will always be changed as `MS_PRIVATE` by runc, but the parent mount of rootfs may have propagation type MS_SHARED.
Before this patch, it will return an error ASAP, but after this patch, it will do some magic operation to let `pivot_root(2)` work, I think we should be careful.

Reply:
> The user case:
> For docker, containerd, kubernetes, the rootfs is always mount as a overlay fs, it will always be changed as `MS_PRIVATE` by runc, but the parent mount of rootfs may have propagation type MS_SHARED.

For this case, this patch has the same security issue like `--no-pivot`, even this flag is not provided. I can escape from the container.
So, I think we should split this PR into two:
1. Deprecate the flag `no-pivot`, this should be in `milestone 1.2.1`;
2. Find a new way to handle initramfs case, this one can be moved to the next version if we can't fix it in a short time.

Reply:
> For this case, this patch has the same security issue like --no-pivot, even this flag is not provided. I can escape from the container.

How? After `pivot_root(2)` the whole host mount tree is gone, no? `chroot_fs_refs` should move the `current->fs->root` to the `pivot_root(2)`'d path so the intermediate `chroot(2)` should not have any effect on the final mount setup AFAICS. Can you share a reproducer?

> For docker, containerd, kubernetes, the rootfs is always mount as a overlay fs, it will always be changed as MS_PRIVATE by runc, but the parent mount of rootfs may have propagation type MS_SHARED.

We also set the parent to `MS_PRVIATE` in `rootfsParentMountPrivate`. (It already should not be the case that we get `-EINVAL` for that reason because we explicitly make sure that it works with `rootfsParentMountPrivate`.) Is that not sufficient?

The bind-mount of `/` keeps the propagation settings the same as well. So if the propagation was actually wrong, the second `pivot_root(2)` would still fail. I also don't see how you could use that to break out of the container -- AFAIK the reason the kernel requires the mount to be `!shared` is so that the mount tree changes during `pivot_root(2)` don't propagate to the host (and break it).

> Find a new way to handle initramfs case, this one can be moved to the next version if we can't fix it in a short time.

The "magic" done is just creating a bind-mount of the root that can be used for `pivot_root(2)` (in an `MS_PRIVATE` mount so the mount itself will be `MS_PRIVATE`, to answer your question).

This is the only way of doing it (I spoke to the VFS maintainers a few weeks ago, and this is the only "right" way of doing it -- FWIW Lennart implied that systemd does this [but they don't it seems](https://github.com/systemd/systemd/blob/v256/src/shared/switch-root.c#L172)).

It's a bit hard to understand what the problem is without an actual reproducer.

Reply:
> We can't deprecate it without having a solution for minikube and kata...

Mark it deprecate not means we remove it right now, it's only to remind users that they should not use it except there is no way to run their case, and to inform users that we can't guarantee  the container's security with `--no-privot`.
But yes, it's a little strange do it in the situation that we still have no solution for minikube and kata or other projects. I think we can find a solution to fix it. I suggest to split into two because maybe we need more time to find out this solution, but I think we should release `v1.2.1` ASAP, then we can let runc `v1.2.*` could be tested by more and more users, especially in some downstream rc versions.

> We also set the parent to `MS_PRVIATE` in `rootfsParentMountPrivate`. (It already should not be the case that we get `-EINVAL` for that reason because we explicitly make sure that it works with `rootfsParentMountPrivate`.) Is that not sufficient?

No. I have concern about it for a long time after #4417 merged, the commit is: 13a6f56097e9d6be75b15ddd33749dc92d1bc88c. This commit has changed the behavior of `rootfsParentMountPrivate`:
1. Before this commit, we make `/` to `MS_PRIVATE` mount;
2. After this commit, if the `rootfs` is a overlay mount, we will only make `rootfs` to `MS_PRIVATE` mount.
@kolyshkin I don't know this behavior change will cause some issues, but we will not set parent to `MS_PRIVATE` is a fact, but maybe make `rootfs` `MS_PRIVATE` is enough.


> It's a bit hard to understand what the problem is without an actual reproducer.

I'll send you a email ASAP.